kvserver: record metrics for ErrProposalDropped

tbg opened the pull request:
Touches https://github.com/cockroachdb/cockroach/issues/100096.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100083)
<!-- Reviewable:end -->

tbg commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -490,7 +497,7 @@ func (b *propBuf) FlushLockedWithRaftGroup(
 			// preserve the correct ordering or proposals. Later proposals
 			// will start a new batch.
-			if err := proposeBatch(raftGroup, b.p.getReplicaID(), ents); err != nil {
+			if err := proposeBatch(ctx, raftGroup, b.p.getReplicaID(), ents, buf[i-len(ents):len(ents)]); err != nil {
```
I tried to simplify the code (which also wasn't correct when you reviewed it anyway). We no longer shift the slices around. Instead, we track a `baseIndex` below which we've already proposed. I think this came out simpler, PTAL.

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -416,12 +418,22 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+	// `baseIdx` is the index into `ents` of the first element that still has
+	// to get flushed.
+	// INVARIANT: ents[i] corresponds to the proposal at buf[i].
+	baseIdx := 0
```
What are we trying to do?

We have the `buf` with the proposals, and `ents` with at most `len(buf)` proposals (some of them can be skipped). For example:

```
buf / props: 1 2 3 4 5 6 7 8 9
       ents: 1 2 4 6 9
```

The `ents` are proposed in chunks, say: `[1 2 4]`, then `[6 9]`.

We want to annotate the `ents` slices with the corresponding `buf / props`, correct? So, we want the iteration to observe:

```
buf / props: 1 2 x 4 | x 6 x x 9
       ents: 1 2   4 |   6     9
```

Instead of writing a fake entry in `ents`, how about "compacting" `props` along the way? (since they are `nil`-ed out anyway)

So:

```golang
buf := b.arr.asSlice()[:used]
ents := make([]raftpb.Entry, 0, used)
defer nilOutBuf()

nextProp := 0 // INVARIANT: len(ents) == nextProp
for _, p := range buf {
	...
	if weWantToAppendToEnts() {
		ents = append(ents, raftpb.Entry{...})
		buf[nextProp] = p
		nextProp++
	}
	if weWantToSendBatch() {
		err := proposeBatch(ctx, raftGroup, b.p.getReplicaID(), ents, buf[:nextProp])
		...
	}
	...
	err := proposeBatch(ctx, raftGroup, b.p.getReplicaID(), ents, buf[:nextProp])
}
```

pavelkalinnikov replied:
There are some buglets in this snippet, I think we need to also keep track of where we start. But you get the idea hopefully: instead of `ifWeDontWantToAppendToEnts() { insert fake Entry{} to ents}`, flip the condition to `ifWeWantToAppendToEnts() { also track props }`

pavelkalinnikov replied:
Also, it might be beneficial to move the invariant maintenance into a `func` next to `buf` and `ents` definition. The loop is quite large, so it's hard to identify the algorithm in it.

tbg replied:
I basically took this suggestion, but I think we can simplify it? We only flush "everything" on ConfChange, so we can "just" keep a pointer to first and last element in `buf` that is current and update. I don't see why we have to mutate the `buf` slice. Chances are I'm asleep at the wheel, but this PR isn't ready for review again anyway, so hold off on checking while I sleep over it and clean up a bit more.

pav-kv reviewed (commented):
Dumping some comments for now. The new version seems confusing too at the first glance. Proposed a simpler solution, if I understood the problem correctly. PTAL.

tbg committed:
kvserver: clear propBuf in one go

Enables a future change.

Epic: none
Release note: None

tbg committed:
kvserver: small code movement

Epic: none
Release note: None

tbg committed:
kvserver: (another) small code movement

Epic: none
Release note: None

tbg committed:
kvserver: pass matching props to proposeBatch

Epic: none
Release note: None

tbg committed:
kvserver: simplify return in proposeBatch

Epic: none
Release note: None

tbg commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -899,7 +920,18 @@ func (b *propBuf) forwardClosedTimestampLocked(closedTS hlc.Timestamp) bool {
+	props []*ProposalData, // must match ents slice
+) error {
+	if len(ents) != len(props) {
+		return errors.AssertionFailedf("ents and props don't match up: %v and %v", ents, props)
```
Maybe - but it's a clear programming error, so I don't want to go crazy and definitely don't want to under-report information, in case the error is rare. I'll leave as is unless you feel strongly.

pavelkalinnikov replied:
I don't know how far up the stack this error can make, whether it can make into text logs or get outside the process. It can contain user data, right? So that's a bit sensitive. Maybe err on the safe side, and include only necessary info in the error message. See a similar `panic` in `maybeDeductFlowTokens`.

tbg commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -912,6 +944,17 @@ func proposeBatch(raftGroup proposerRaft, replID roachpb.ReplicaID, ents []raftp
 		// ignored prior to the introduction of ErrProposalDropped).
```
It's still true, better handling of it would at least entail adding metrics, etc.
I think other than that it should already work about as well as it can, we'll repropose everything when we get a `reasonNewLeaderOrConfChange` refresh, but I can't say this has exactly been verified 😄 so I think this TODO should stay for now.

tbg committed:
kvserver: log to trace on proposal drop

Epic: none
Release note: None

tbg committed:
kvserver: add proposer.onErrProposalDropped

With noop impl for now.

Epic: none
Release note: None

tbg committed:
kvserver: plumb admitHandles into proposeBatch

One step to removing repetitive handling at the caller.

Epic: none
Release note: None

tbg committed:
kvserver: split up proposer interface

We'll pass the small part into `proposeBatch`.

Epic: none
Release note: None

tbg committed:
kvserver: move call to maybeDeductFlowTokens into proposeBatch

Epic: none
Release note: None

tbg committed:
kvserver: invoke onErrProposalDropped

The impl is still a no-op, so this change is a no-op, too.

Epic: none
Release note: None

tbg committed:
kvserver: add raft.dropped for ErrProposalsDropped

This counts the number of entries raft drops on the floor. There are two reasons for this:

- uncommitted log size exceeded
- no leader known

They're not currently discriminated. Either way, we want neither of these to
occur, but we know from experience that they do. This metric will give us a
sense of the frequency and allow us to more conclusively correlate with splits,
etc.

Epic: none
Release note: None

tbg commented:
Starting to pick this back up, no need to re-review yet so I converted to draft. I'll reach out when it's ready.

tbg committed:
kvserver: get replicaID from p in proposeBatch

Epic: none
Release note: None

tbg committed:
kvserver: add `raft.dropped_leader`

This allows us to discriminate between MsgProp that are dropped because no
leader is known vs those for which the leader is unwilling to accept more
proposals (uncommitted log size being the relevant reason today[^1]).

[^1]: raft will also drop conf changes if a pending conf change is still
pending, but since our conf changes are raft-inflight exactly when the
RangeDescriptor on the Replica has an intent, we don't hit this case in CRDB.

Epic: none
Release note: none

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -588,7 +588,7 @@ func (b *propBuf) FlushLockedWithRaftGroup(
 	}
 
-	propErr, _ := proposeBatch(ctx, b.p, raftGroup, b.p.getReplicaID(), ents, admitHandles, buf[firstProp:nextProp])
+	propErr, _ := proposeBatch(ctx, b.p, raftGroup, ents, admitHandles, buf[firstProp:nextProp])
```
Do any of the callers still need the second returned parameter?
Tests, potentially? But I don't see any. Consider removing it then.

pavelkalinnikov replied:
I don't know how far up the stack this error can make, whether it can make into text logs or get outside the process. It can contain user data, right? So that's a bit sensitive. Maybe err on the safe side, and include only necessary info in the error message. See a similar `panic` in `maybeDeductFlowTokens`.

pavelkalinnikov commented on pkg/kv/kvserver/metrics.go:
```diff
@@ -1090,6 +1090,13 @@ or the delegate being too busy to send.
+	metaRaftProposalsDroppedLeader = metric.Metadata{
+		Name: "raft.dropped_leader",
+		Help: "Number of Raft proposals dropped by a Replica that believes itself to be the leader " +
+			"(this counts individial raftpb.Entry, not raftpb.MsgProp)",
```
Could add a note that this is a "subset" of the `raft.dropped` metric.

pav-kv reviewed (approved):
Super clean PR, thanks.

tbg committed:
kvserver: remove dropped return param in proposeBatch

tbg commented:
bors r=pavelkalinnikov
TFTR! Addressed your comments in the affirmative.

tbg commented:
bors r-

tbg commented:
bors r=pavelkalinnikov

Removing the `dropped` param set off the linter since it now understood an unhandled error it didn't before.
